# Phase 8B — Higher-Order Temporal Structure

## Exact three-step and conditional four-step sequence analysis

Phase 8B asks whether longer ordered displacement sequences contain
forward-versus-reversed information beyond net directional displacement.

The analysis design was pre-specified and committed before inspecting
Phase 8B outcome statistics.

Primary sequence length:

`3 steps`

Conditional secondary sequence length:

`4 steps`

The first data-dependent task is deliberately limited to a
**sample-support audit**.

At this stage we will determine only:

- the number of exact three-step sequences;
- the number of trajectory groups contributing to them;
- the number of exact four-step sequences;
- the number of trajectory groups contributing to them;
- whether the pre-specified four-step support gate is satisfied.

No classifier, AUC, p-value, null simulation, or irreversibility statistic
is computed in this section.

The pre-specified four-step gate is:

`at least 200 exact sequences AND at least 12 trajectory groups`.

In [1]:
import numpy as np
import pandas as pd

from _path import PROJECT_ROOT

from src.steps import validate_tracks
from src.sequences import build_k_step_sequences


CTC_REFERENCE_PATH = (
    PROJECT_ROOT
    / "data"
    / "MSC01_tracks_ctc_reference.csv"
)

TRACKMATE_RAW_PATH = (
    PROJECT_ROOT
    / "data"
    / "MSC01_tracks_clean.csv"
)

TRACKMATE_COMOVING_PATH = (
    PROJECT_ROOT
    / "data"
    / "MSC01_tracks_comoving.csv"
)


print("Project root:", PROJECT_ROOT)

print(
    "CTC reference exists:",
    CTC_REFERENCE_PATH.exists(),
)

print(
    "TrackMate raw exists:",
    TRACKMATE_RAW_PATH.exists(),
)

print(
    "TrackMate co-moving exists:",
    TRACKMATE_COMOVING_PATH.exists(),
)

Project root: C:\Users\Abolfazl.PH\Desktop\cell-irreversibility
CTC reference exists: True
TrackMate raw exists: True
TrackMate co-moving exists: True


In [2]:
ctc_reference_table = pd.read_csv(
    CTC_REFERENCE_PATH
)

trackmate_raw = pd.read_csv(
    TRACKMATE_RAW_PATH
)

trackmate_comoving = pd.read_csv(
    TRACKMATE_COMOVING_PATH
)


ctc_raw = (
    ctc_reference_table[
        [
            "track_id",
            "frame",
            "time_min",
            "x_um",
            "y_um",
        ]
    ]
    .rename(
        columns={
            "track_id": "cell_id",
            "time_min": "t_min",
        }
    )
    .sort_values(
        [
            "cell_id",
            "frame",
        ]
    )
    .reset_index(
        drop=True
    )
)


trackmate_raw = (
    trackmate_raw[
        [
            "cell_id",
            "frame",
            "t_min",
            "x_um",
            "y_um",
        ]
    ]
    .sort_values(
        [
            "cell_id",
            "frame",
        ]
    )
    .reset_index(
        drop=True
    )
)


trackmate_comoving = (
    trackmate_comoving[
        [
            "cell_id",
            "frame",
            "t_min",
            "x_um",
            "y_um",
        ]
    ]
    .sort_values(
        [
            "cell_id",
            "frame",
        ]
    )
    .reset_index(
        drop=True
    )
)


print(
    "CTC shape:",
    ctc_raw.shape,
)

print(
    "CTC tracks:",
    ctc_raw["cell_id"].nunique(),
)

print(
    "TrackMate raw shape:",
    trackmate_raw.shape,
)

print(
    "TrackMate raw tracks:",
    trackmate_raw["cell_id"].nunique(),
)

print(
    "TrackMate co-moving shape:",
    trackmate_comoving.shape,
)

print(
    "TrackMate co-moving tracks:",
    trackmate_comoving["cell_id"].nunique(),
)

CTC shape: (428, 5)
CTC tracks: 15
TrackMate raw shape: (474, 5)
TrackMate raw tracks: 35
TrackMate co-moving shape: (474, 5)
TrackMate co-moving tracks: 35


In [3]:
validate_tracks(
    ctc_raw
)

validate_tracks(
    trackmate_raw
)

validate_tracks(
    trackmate_comoving
)


trackmate_raw_structure = (
    trackmate_raw[
        [
            "cell_id",
            "frame",
            "t_min",
        ]
    ]
)

trackmate_comoving_structure = (
    trackmate_comoving[
        [
            "cell_id",
            "frame",
            "t_min",
        ]
    ]
)


trackmate_structure_identical = (
    trackmate_raw_structure.equals(
        trackmate_comoving_structure
    )
)


print(
    "TrackMate raw/co-moving structure identical:",
    trackmate_structure_identical,
)


assert trackmate_structure_identical

assert len(ctc_raw) == 428
assert ctc_raw["cell_id"].nunique() == 15

assert len(trackmate_raw) == 474
assert trackmate_raw["cell_id"].nunique() == 35


print(
    "Phase 8B trajectory structural validation: PASSED"
)

TrackMate raw/co-moving structure identical: True
Phase 8B trajectory structural validation: PASSED


In [4]:
ctc_three_step_raw = (
    build_k_step_sequences(
        ctc_raw,
        n_steps=3,
    )
)

ctc_four_step_raw = (
    build_k_step_sequences(
        ctc_raw,
        n_steps=4,
    )
)


trackmate_three_step_raw = (
    build_k_step_sequences(
        trackmate_raw,
        n_steps=3,
    )
)

trackmate_four_step_raw = (
    build_k_step_sequences(
        trackmate_raw,
        n_steps=4,
    )
)


print(
    "CTC exact 3-step sequences:",
    len(ctc_three_step_raw),
)

print(
    "CTC 3-step groups:",
    ctc_three_step_raw[
        "cell_id"
    ].nunique(),
)

print()

print(
    "CTC exact 4-step sequences:",
    len(ctc_four_step_raw),
)

print(
    "CTC 4-step groups:",
    ctc_four_step_raw[
        "cell_id"
    ].nunique(),
)

print()

print(
    "TrackMate exact 3-step sequences:",
    len(trackmate_three_step_raw),
)

print(
    "TrackMate 3-step groups:",
    trackmate_three_step_raw[
        "cell_id"
    ].nunique(),
)

print()

print(
    "TrackMate exact 4-step sequences:",
    len(trackmate_four_step_raw),
)

print(
    "TrackMate 4-step groups:",
    trackmate_four_step_raw[
        "cell_id"
    ].nunique(),
)

CTC exact 3-step sequences: 383
CTC 3-step groups: 14

CTC exact 4-step sequences: 369
CTC 4-step groups: 14

TrackMate exact 3-step sequences: 357
TrackMate 3-step groups: 28

TrackMate exact 4-step sequences: 329
TrackMate 4-step groups: 26


In [5]:
for (
    label,
    sequences,
    expected_steps,
) in [
    (
        "CTC 3-step",
        ctc_three_step_raw,
        3,
    ),
    (
        "CTC 4-step",
        ctc_four_step_raw,
        4,
    ),
    (
        "TrackMate 3-step",
        trackmate_three_step_raw,
        3,
    ),
    (
        "TrackMate 4-step",
        trackmate_four_step_raw,
        4,
    ),
]:

    assert (
        sequences[
            "sequence_span_frames"
        ]
        == expected_steps
    ).all()

    assert np.allclose(
        sequences[
            "sequence_span_min"
        ],
        expected_steps * 20.0,
    )

    assert (
        sequences[
            "frame_end"
        ]
        - sequences[
            "frame_start"
        ]
        == expected_steps
    ).all()


print(
    "Exact 3-step/4-step temporal-span validation: PASSED"
)

Exact 3-step/4-step temporal-span validation: PASSED


In [6]:
FOUR_STEP_MIN_SEQUENCES = 200
FOUR_STEP_MIN_GROUPS = 12


support_by_source = pd.DataFrame(
    [
        {
            "tracking_source":
                "CTC reference",

            "n_three_step_sequences":
                len(
                    ctc_three_step_raw
                ),

            "n_three_step_groups":
                ctc_three_step_raw[
                    "cell_id"
                ].nunique(),

            "n_four_step_sequences":
                len(
                    ctc_four_step_raw
                ),

            "n_four_step_groups":
                ctc_four_step_raw[
                    "cell_id"
                ].nunique(),
        },
        {
            "tracking_source":
                "TrackMate",

            "n_three_step_sequences":
                len(
                    trackmate_three_step_raw
                ),

            "n_three_step_groups":
                trackmate_three_step_raw[
                    "cell_id"
                ].nunique(),

            "n_four_step_sequences":
                len(
                    trackmate_four_step_raw
                ),

            "n_four_step_groups":
                trackmate_four_step_raw[
                    "cell_id"
                ].nunique(),
        },
    ]
)


support_by_source[
    "four_step_eligible"
] = (
    (
        support_by_source[
            "n_four_step_sequences"
        ]
        >= FOUR_STEP_MIN_SEQUENCES
    )
    &
    (
        support_by_source[
            "n_four_step_groups"
        ]
        >= FOUR_STEP_MIN_GROUPS
    )
)


support_audit_records = []

for _, row in (
    support_by_source.iterrows()
):

    for representation in [
        "raw",
        "co-moving",
    ]:

        support_audit_records.append(
            {
                "tracking_source":
                    row[
                        "tracking_source"
                    ],

                "representation":
                    representation,

                "n_three_step_sequences":
                    int(
                        row[
                            "n_three_step_sequences"
                        ]
                    ),

                "n_three_step_groups":
                    int(
                        row[
                            "n_three_step_groups"
                        ]
                    ),

                "n_four_step_sequences":
                    int(
                        row[
                            "n_four_step_sequences"
                        ]
                    ),

                "n_four_step_groups":
                    int(
                        row[
                            "n_four_step_groups"
                        ]
                    ),

                "four_step_eligible":
                    bool(
                        row[
                            "four_step_eligible"
                        ]
                    ),
            }
        )


phase8b_support_audit = (
    pd.DataFrame(
        support_audit_records
    )
)


display(
    phase8b_support_audit
)


print(
    "\nFour-step gate:"
)

print(
    "minimum sequences =",
    FOUR_STEP_MIN_SEQUENCES,
)

print(
    "minimum groups =",
    FOUR_STEP_MIN_GROUPS,
)


print(
    "\nPhase 8B sample-support audit: COMPLETE"
)

,tracking_source,representation,n_three_step_sequences,n_three_step_groups,n_four_step_sequences,n_four_step_groups,four_step_eligible
0,CTC reference,raw,383,14,369,14,True
1,CTC reference,co-moving,383,14,369,14,True
2,TrackMate,raw,357,28,329,26,True
3,TrackMate,co-moving,357,28,329,26,True



Four-step gate:
minimum sequences = 200
minimum groups = 12

Phase 8B sample-support audit: COMPLETE


## Phase 8B.1 — Exact sequence-support audit

The first data-dependent Phase 8B computation was restricted to temporal
sample support, as required by the pre-specified analysis design.

No classifier performance, null statistic, p-value, or irreversibility
quantity was inspected.

### Three-step support

CTC reference tracking produced:

- 383 exact three-step sequences;
- 14 contributing reference tracks.

TrackMate produced:

- 357 exact three-step sequences;
- 28 contributing tracks.

### Four-step support

CTC reference tracking produced:

- 369 exact four-step sequences;
- 14 contributing reference tracks.

TrackMate produced:

- 329 exact four-step sequences;
- 26 contributing tracks.

The pre-specified four-step eligibility rule was:

`at least 200 exact four-step sequences AND at least 12 trajectory groups`.

Both tracking sources satisfy this rule.

Therefore:

- three-step analysis remains the primary Phase 8B analysis;
- four-step analysis is formally eligible and will be retained as the
  pre-specified secondary extension;
- the four-step support thresholds will not be modified later in response to
  inferential results.

Raw and co-moving representations have identical temporal support because
common-motion subtraction changes trajectory coordinates but does not add or
remove track-frame observations.

The support audit is now closed.

In [7]:
ctc_four_step_eligible = bool(
    support_by_source.loc[
        support_by_source[
            "tracking_source"
        ].eq("CTC reference"),
        "four_step_eligible",
    ].iloc[0]
)

trackmate_four_step_eligible = bool(
    support_by_source.loc[
        support_by_source[
            "tracking_source"
        ].eq("TrackMate"),
        "four_step_eligible",
    ].iloc[0]
)


assert ctc_four_step_eligible
assert trackmate_four_step_eligible


print(
    "CTC four-step eligibility:",
    ctc_four_step_eligible,
)

print(
    "TrackMate four-step eligibility:",
    trackmate_four_step_eligible,
)

print(
    "Phase 8B support-gate decision: RECORDED"
)

CTC four-step eligibility: True
TrackMate four-step eligibility: True
Phase 8B support-gate decision: RECORDED
